### PR TITLE
Uptime tests poller group

### PR DESCRIPTION
**Pull request by tbuchaillot**
## Description

Added new configuration for uptime tests: `poller_group` - string - Adds a postfix to the Poller Cache key. In that way, we can have different pollers for each group.

An example of a config with `poller_group`:
```
 "uptime_tests": {
        "disable": false,
        "poller_group":"external",
        "config": {
            "failure_trigger_sample_size": 1,
            "time_wait": 2,
            "checker_pool_size": 50,
            "enable_uptime_analytics": true
        }
    },
```

Modified `uptime_tests.disable` to be the trigger of the HostCheckerManager. If it is true, the gateway won't be able to become the Host checker manager.
Added some host checker manager unit tests.

## Related Issue

https://tyktech.atlassian.net/browse/TT-496

## Motivation and Context

On sharded gateway environments, each shard / node should have their own Poller / Host checker to do the uptimetest of their APIs. For more context, please check https://tyktech.atlassian.net/browse/TT-496

## How This Has Been Tested

Tested with:
2 GWs: poller_group:"a" , db_app_conf_options.node_is_segmented:true and db_app_conf_options.tags:["test"]
1 GWs: poller_group:"b" , db_app_conf_options.node_is_segmented:true and db_app_conf_options.tags:["test-2"]
1 Pump running with uptime tests configured.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

File: gateway/host_checker_manager_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	"text/template"
+
+	"github.com/TykTechnologies/tyk/config"
+	"github.com/TykTechnologies/tyk/storage"
+	uuid "github.com/satori/go.uuid"
+)
+
+func TestHostCheckerManagerInit(t *testing.T) {
+	ts := StartTest()
+	defer ts.Close()
```
Comment (gernest) [inline gateway/host_checker_manager_test.go]:
```suggestion
```

Why start this and not use it at all. Redis will be up and running already before `Main` function. You can do redis stuff fine.

**Review by matiasinsaurralde — APPROVED** [review]

File: gateway/host_checker_manager_test.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+	activeInstance, err = hc.store.GetKey("PollerActiveInstanceID")
+	if err != nil {
+		t.Error("PollerActiveInstanceID should exist in redis.", activeInstance)
+	}
+	if activeInstance != hc.Id {
+		t.Error("PollerActiveInstanceID value should be hc.Id")
+	}
+
+}
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:85]: We are hardcoding `PollerActiveInstanceID` and also the errors  lack  more context.

We should have something like 

```diff
diff --git a/gateway/host_checker_manager_test.go b/gateway/host_checker_manager_test.go
index c320147e..cb1dca59 100644
--- a/gateway/host_checker_manager_test.go
+++ b/gateway/host_checker_manager_test.go
@@ -38,7 +38,8 @@ func TestAmIPolling(t *testing.T) {
 
 	//Testing if we had 2 active host checker managers, only 1 takes control of the uptimechecks
 	globalConf := config.Global()
-	globalConf.UptimeTests.PollerGroup = "TEST"
+	groupID := "TEST"
+	globalConf.UptimeTests.PollerGroup = groupID
 	config.SetGlobal(globalConf)
 	defer ResetTestConfig()
 
@@ -58,12 +59,13 @@ func TestAmIPolling(t *testing.T) {
 	}
 
 	//Testing if the PollerCacheKey contains the poller_group
-	activeInstance, err := hc.store.GetKey("PollerActiveInstanceID.TEST")
+	testKey := PollerCacheKey + "." + groupID
+	activeInstance, err := hc.store.GetKey(testKey)
 	if err != nil {
-		t.Error("PollerActiveInstanceID.TEST  should exist in redis.", activeInstance)
+		t.Errorf("%q  should exist in redis.%v", testKey, activeInstance)
 	}
 	if activeInstance != hc.Id {
-		t.Error("PollerActiveInstanceID.TEST value should be hc.Id")
+		t.Errorf("%q value : expected %v got %v", testKey, hc.Id, activeInstance)
 	}
 
 	ResetTestConfig()
@@ -74,12 +76,12 @@ func TestAmIPolling(t *testing.T) {
 	hc.AmIPolling()
 	defer hc.StopPoller()
 
-	activeInstance, err = hc.store.GetKey("PollerActiveInstanceID")
+	activeInstance, err = hc.store.GetKey(PollerCacheKey)
 	if err != nil {
-		t.Error("PollerActiveInstanceID should exist in redis.", activeInstance)
+		t.Errorf("%q should exist in redis.%v", PollerCacheKey, activeInstance)
 	}
 	if activeInstance != hc.Id {
-		t.Error("PollerActiveInstanceID value should be hc.Id")
+		t.Errorf("%q : value expected %v got %v", PollerCacheKey, hc.Id, activeInstance)
 	}
 
 }
```

File: gateway/host_checker_manager_test.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+	ctx := context.TODO()
+	go hc.CheckActivePollerLoop(ctx)
+	defer hc.StopPoller()
+
+	found := false
+
+	//Giving 5 retries to find the poller active key
+	for i := 0; i < 5; i++ {
+		activeInstance, err := hc.store.GetKey("PollerActiveInstanceID")
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:114]:
```suggestion
		activeInstance, err := hc.store.GetKey(PollerCacheKey)
```

File: gateway/host_checker_manager_test.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		if activeInstance == hc.Id && err == nil {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		t.Error("activeInstance should be hc.Id when the CheckActivePollerLoop is running")
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:122]:
```suggestion
		t.Errorf("activeInstance should be %q when the CheckActivePollerLoop is running", hc.Id)
```

File: gateway/host_checker_manager_test.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	config.SetGlobal(globalConf)
+	defer ResetTestConfig()
+
+	redisStorage := &storage.RedisCluster{KeyPrefix: "host-checker-test:"}
+	hc.Init(redisStorage)
+	hc2 := HostCheckerManager{}
+	hc2.Init(redisStorage)
+
+	polling = hc.AmIPolling()
+	defer hc.StopPoller()
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:51]: `hc.StartPoller` is not called why do this then?

File: gateway/host_checker_manager_test.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+	pollingHc2 := hc2.AmIPolling()
+	defer hc2.StopPoller()
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:54]: `hc2.StartPoller` is not called why do this then?

File: gateway/host_checker_manager_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		t.Error("PollerActiveInstanceID.TEST value should be hc.Id")
+	}
+
+	ResetTestConfig()
+	//Testing if the PollerCacheKey doesn't contains the poller_group by default
+	hc = HostCheckerManager{}
+	redisStorage = &storage.RedisCluster{KeyPrefix: "host-checker-test:"}
+	hc.Init(redisStorage)
+	hc.AmIPolling()
+	defer hc.StopPoller()
```
Comment (gernest) [inline gateway/host_checker_manager_test.go:75]: `hc.StartPoller` is not called why do this then?

**Review by gernest — DISMISSED** [review]
Just some suggestions.

I am having a hard time since I can't fully run all the tests locally to reproduce this. 
Maybe address the suggestions then try to do something like this.

- Skip all the new tests , and push changes.  If the tests pass then..
- Start enabling one new test at a time and push changes (observe ci) until we find the offending tests

If you try first step and the same issue is observed we can take it from there.

**Comment by gernest** [comment]
@tbuchaillot   Please let's not merge this until all tests are passing. I have a feeling something is missing on implementation part and not the new tests.

I don't want to hijack this work so I will leave you to `bisect` the offending parts.

**Review by matiasinsaurralde — APPROVED** [review]

**Review by sredxny — APPROVED** [review]
LGTM

**Pull request merged**